Location path 2

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix

**What is the current behavior? (You can also link to an open issue here)**
see the commit message

**What is the new behavior (if this is a feature change)?**
see the commit message

**Does this PR introduce a breaking change?**
The previous behaviour would have led to a broken app, so it is highly
unlikely that any app relied on this behaviour. So this doesn't constitute
a breaking change.

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

Alternative solution #15359

petebacondarwin committed:
test(): remove redundant hashPrefix params

petebacondarwin commented on src/ng/location.js:
```diff
@@ -144,9 +149,10 @@ function LocationHtml5Url(appBase, appBaseNoFile, basePrefix) {
     if (isDefined(appUrl = stripBaseUrl(appBase, url))) {
       prevAppUrl = appUrl;
-      if (isDefined(appUrl = stripBaseUrl(basePrefix, appUrl))) {
+      if (basePrefix && isDefined(appUrl = stripBaseUrl(basePrefix, appUrl))) {
```
interestingly I think this was actually the cause of the bug in the first place...

mgol replied:
There are no tests for this change, are there?

petebacondarwin replied:
No new tests, specifically, but it doesn't break any of the current tests
And there are the new tests with the double slashes.
I couldn't think of a specific test for this change. Any ideas?

koto commented on test/ng/locationSpec.js:
```diff
@@ -2480,6 +2480,15 @@ describe('$location', function() {
+      }).toThrowMinErr('$location', 'badpath');
+
+      expect(function() {
+        parseLinkAndReturn(locationUrl, 'http://server/pre/\\\\other/path');
```
Could you add a test case with mixed slashes (`pre//\\//other/path`) ?

petebacondarwin replied:
will do

gkalpak commented on src/ng/location.js:
```diff
@@ -30,13 +30,18 @@ function parseAbsoluteUrl(absoluteUrl, locationObj) {
 
-function parseAppUrl(relativeUrl, locationObj) {
-  var prefixed = (relativeUrl.charAt(0) !== '/');
+  if (url.search(DOUBLE_SLASH_REGEX) === 0) {
```
<sub>You could also add `^` at the beginning of `DOUBLE_SLASH_REGEX` and use  `DOUBLE_SLASH_REGEX.test(url)`?</sub>

petebacondarwin committed:
fix($location): throw if the path starts with double (back)slashes

Previously `$location` was rewriting such paths to remove not only the
double slashes but also the first segment of the path, leading to an invalid
path.

In this change, we deem leading double (back)slashes an invalid path and
now throw a `$location:badpath` error if that occurs.

Closes #15365

petebacondarwin commented:
@koto and @mprobst - any further comments before we merge this?

koto commented:
Looks good to me, security-wise.

On Mon, Nov 7, 2016, 12:10 Pete Bacon Darwin <notifications@github.com>
wrote:

> @koto <https://github.com/koto> and @mprobst <https://github.com/mprobst>
> - any further comments before we merge this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/15365#issuecomment-258808297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAH0q3zX-T7nINjGyNQnmgFsrH2up6INks5q7wc6gaJpZM4KpoUt>
> .
>

mprobst commented:
LGTM

ivanrey commented:
This fix has problems with internal routing of an app. Sometimes double slash in fragments might be important to indicate an specific state.

I believe this fix should not take into account the fragment part of the URL since it's not sent to the server and breaks some applications.

gkalpak commented:
@ivanrey, afaict this is supposed to only check for the beginning of the client-side URL and I think it didn't work correctly before either. Do you have a specific usecase/demo you can share, that used to work before and broke as a result of this fix?
